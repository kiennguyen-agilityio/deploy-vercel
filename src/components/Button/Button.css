.btn-primary {
  background: var(--color-black);
  border-radius: 10px;
  color: var(--color-white);
  display: flex;
  align-items: center;
  justify-content: center;
  font: normal var(--font-weight-semibold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
}

.btn-secondary {
  border: 1px solid var(--color-black);
  color: var(--color-black);
  background-color: var(--color-white);
  border-radius: 10px;
  display: flex;
  align-items: center;
  justify-content: center;
  font: normal var(--font-weight-semibold) var(--font-size-sm) / var(--line-height-md) 'Open Sans';
}

.btn.btn-tertiary {
  border: none;
  background: none;
  width: 9px;
  height: 34px;
  padding: 0px;
  margin: 20px;
  font-size: var(--font-size-l);
}

.btn.btn-tertiary.btn-color-warning,
.btn.btn-tertiary.btn-color-success {
  color: var(--color-button-secondary);
}

.btn.btn-tertiary.btn-color-warning,
.btn.btn-loading.btn-color-warning {
  color: var(--color-button-primary);
}

.btn-loading {
  background-color: var();
  color: var();
}

button[disabled] {
  border: 1px solid var();
  background-color: var();
  color: var();
}

button[disabled]:hover {
  background-color: var();
  color: var();
  cursor: no-drop;
}

.btn img {
  width: 100%;
}

.btn-sm {
  width: 219px;
  height: 60px;
}

.btn-lg {
  width: 254px;
  height: 60px;
}

.btn-xl {
  width: 284px;
  height: 60px;
}
