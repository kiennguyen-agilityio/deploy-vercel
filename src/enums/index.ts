export enum BUTTON_VARIANTS {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  TERTIARY = 'tertiary',
}

export enum BUTTON_SIZES {
  SMALL = 'sm',
  MEDIUM = 'md',
  LARGE = 'lg',
  EXTRA_LARGE = 'xl',
}

export enum BUTTON_COLORS {
  SUCCESS = 'success',
  WARNING = 'warning',
  DEFAULT = 'default',
}

export enum BUTTON_TYPES {
  BUTTON = 'button',
  SUBMIT = 'submit',
  RESET = 'reset',
}

export enum SORT {
  ASC = 'asc',
  DESC = 'desc',
}
