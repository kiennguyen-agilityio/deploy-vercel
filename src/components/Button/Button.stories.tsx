import type { Meta, StoryObj } from '@storybook/react';

import { BUTTON_VARIANTS } from '@/enums/index';

import Button from './Button';

const meta: Meta<typeof Button> = {
  component: Button,
};

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    label: 'Button',
    variant: BUTTON_VARIANTS.PRIMARY,
  },
};

export const Secondary: Story = {
  render: () => <Button label="Button" variant={BUTTON_VARIANTS.SECONDARY} />,
};

export default meta;
