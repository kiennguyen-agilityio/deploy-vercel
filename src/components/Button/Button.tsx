import { ReactElement, ReactNode, memo } from 'react';
import { BUTTON_VARIANTS, BUTTON_SIZES, BUTTON_COLORS, BUTTON_TYPES } from '@/enums/index';

// style
import './Button.css';

export interface Props {
  isDisabled?: boolean;
  isLoading?: boolean;
  label: string;
  leftIcon?: ReactNode;
  rightIcon?: ReactNode;
  variant: BUTTON_VARIANTS;
  size?: BUTTON_SIZES;
  color?: BUTTON_COLORS;
  type?: BUTTON_TYPES;
  onClick?: () => void;
}

const Button = ({
  isDisabled = false,
  isLoading = false,
  label,
  leftIcon,
  rightIcon,
  variant = BUTTON_VARIANTS.PRIMARY,
  size = BUTTON_SIZES.SMALL,
  color = BUTTON_COLORS.DEFAULT,
  type = BUTTON_TYPES.BUTTON,
  onClick,
}: Props): ReactElement => {
  const classes = `btn btn-${variant} btn-color-${color} btn-${size} ${
    isDisabled ? 'btn-disabled' : ''
  } ${isLoading ? 'btn-loading' : ''}`;

  return (
    <button type={type} className={classes} onClick={onClick} disabled={isDisabled || isLoading}>
      {isLoading && <span className="loader" />}
      {leftIcon}
      <span>{label}</span>
      {rightIcon}
    </button>
  );
};

export default memo(Button);
