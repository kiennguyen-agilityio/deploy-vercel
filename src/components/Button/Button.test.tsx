import { render, screen, fireEvent } from '@testing-library/react';
import Button, { Props } from './Button';
import { BUTTON_VARIANTS } from '@/enums/index';
const setup = (props = {}) => {
  const defaultProps: Props = {
    label: 'Submit',
    variant: BUTTON_VARIANTS.PRIMARY,
    onClick: jest.fn(),
    isLoading: false,
    leftIcon: null,
    ...props,
  };

  render(<Button {...defaultProps} />);
};

describe('Button component', () => {
  test('renders button correctly', () => {
    setup();
    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  test('renders button with secondary variant', () => {
    setup({ variant: 'secondary' });

    const buttonElement = screen.getByText('Submit');
    expect(buttonElement).toMatchSnapshot();
  });

  test('calls onClick handler when clicked', () => {
    const handleClick = jest.fn();
    setup({ onClick: handleClick });

    const buttonElement = screen.getByText('Submit');
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('renders button with loading spinner when isLoading is true', () => {
    setup({ isLoading: true });

    const loaderElement = screen.getByRole('button');
    expect(loaderElement).toBeInTheDocument();
  });

  test('renders button with left icon', () => {
    const leftIcon = <i className="fa fa-arrow-left" data-testid="left-icon" />;
    setup({ leftIcon });

    const iconElement = screen.getByTestId('left-icon');
    expect(iconElement).toMatchSnapshot();
  });
});
