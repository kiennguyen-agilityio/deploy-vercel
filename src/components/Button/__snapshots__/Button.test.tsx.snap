// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders button correctly 1`] = `
<span>
  Submit
</span>
`;

exports[`Button component renders button with left icon 1`] = `
<i
  class="fa fa-arrow-left"
  data-testid="left-icon"
/>
`;

exports[`Button component renders button with secondary variant 1`] = `
<span>
  Submit
</span>
`;
